indexing-service/src/main/java/org/apache/druid/indexing/common/task/CompactionTask.java

```diff
@@ -830,112 +814,34 @@ private static DimensionSchema createDimensionSchema(
+    PartitionConfigurationManager(@Nullable IndexTuningConfig tuningConfig)
     {
-      this.targetCompactionSizeBytes = getValidTargetCompactionSizeBytes(targetCompactionSizeBytes, tuningConfig);
       this.tuningConfig = tuningConfig;
     }
 
     @Nullable
-    IndexTuningConfig computeTuningConfig(List<Pair<QueryableIndex, DataSegment>> queryableIndexAndSegments)
+    IndexTuningConfig computeTuningConfig()
     {
-      if (!hasPartitionConfig(tuningConfig)) {
-        final long nonNullTargetCompactionSizeBytes = Preconditions.checkNotNull(
-            targetCompactionSizeBytes,
-            "targetCompactionSizeBytes"
-        );
-        // Find IndexTuningConfig.maxRowsPerSegment which is the number of rows per segment.
-        // Assume that the segment size is proportional to the number of rows. We can improve this later.
-        final long totalNumRows = queryableIndexAndSegments
-            .stream()
-            .mapToLong(queryableIndexAndDataSegment -> queryableIndexAndDataSegment.lhs.getNumRows())
-            .sum();
-        final long totalSizeBytes = queryableIndexAndSegments
-            .stream()
-            .mapToLong(queryableIndexAndDataSegment -> queryableIndexAndDataSegment.rhs.getSize())
-            .sum();
-
-        if (totalSizeBytes == 0L) {
-          throw new ISE("Total input segment size is 0 byte");
-        }
-
-        final double avgRowsPerByte = totalNumRows / (double) totalSizeBytes;
-        final long maxRowsPerSegmentLong = Math.round(avgRowsPerByte * nonNullTargetCompactionSizeBytes);
-        final int maxRowsPerSegment = Numbers.toIntExact(
-            maxRowsPerSegmentLong,
-            StringUtils.format(
-                "Estimated maxRowsPerSegment[%s] is out of integer value range. "
-                + "Please consider reducing targetCompactionSizeBytes[%s].",
-                maxRowsPerSegmentLong,
-                targetCompactionSizeBytes
+      // Setting maxTotalRows to Long.MAX_VALUE to respect the computed maxRowsPerSegment.
+      // If this is set to something too small, compactionTask can generate small segments
+      // which need to be compacted again, which in turn making auto compaction stuck in the same interval.
+      IndexTuningConfig newTuningConfig = tuningConfig == null
+                                          ? IndexTuningConfig.createDefault()
+                                          : tuningConfig;
+      newTuningConfig = newTuningConfig.withPartitionsSpec(newTuningConfig.getGivenOrDefaultPartitionsSpec());
+      if (newTuningConfig.getPartitionsSpec() instanceof DynamicPartitionsSpec) {
+        final DynamicPartitionsSpec dynamicPartitionsSpec = (DynamicPartitionsSpec) newTuningConfig.getPartitionsSpec();
+        return newTuningConfig.withPartitionsSpec(
+            new DynamicPartitionsSpec(
+                dynamicPartitionsSpec.getMaxRowsPerSegment(),
+                dynamicPartitionsSpec.getMaxTotalRowsOr(Long.MAX_VALUE)
```

Comment:
Alternative that allocates one fewer `TuningConfig` for the `DynamicPartitionsSpec` case:

```java
PartitionsSpec partitionsSpec = newTuningConfig.getGivenOrDefaultPartitionsSpec());
if (partitionsSpec instanceof DynamicPartitionsSpec) {
    final DynamicPartitionsSpec dynamicPartitionsSpec = (DynamicPartitionsSpec) partitionsSpec;
    partitionsSpec = new DynamicPartitionsSpec(
        dynamicPartitionsSpec.getMaxRowsPerSegment(),
        dynamicPartitionsSpec.getMaxTotalRowsOr(Long.MAX_VALUE),
    );
}
return newTuningConfig.withPartitionsSpec(partitionsSpec);
```

Reply:
Changed.